### PR TITLE
fix: pause lifecyle changes and add drained status

### PR DESCRIPTION
**Pull request by kohlisid**
fixes https://github.com/numaproj/numaflow/issues/2027
fixes https://github.com/numaproj/numaflow/issues/1991

DrainedOnPause field is added here
1) It will be set when the pipeline is moved to a `Paused` state after a successful drain has been completed
2) It will always be reset whenever phase is moved back to running

### CODE REVIEW

**Comment by kohlisid** [comment]
1) Pipeline created before ISB Service, and desiredPhase=Paused
```
skohli@macos-JQWR9T560R numaflow % kubectl get pl
NAME              PHASE    VERTICES   AGE   MESSAGE
simple-pipeline   Failed   3          13s   ISB Service not found.

```
We keep the phase as failed and do not try to pause 
https://github.com/numaproj/numaflow/issues/1992#issuecomment-2310697806

**Comment by kohlisid** [comment]
```
 drainedOnPause: true
  lastUpdated: "2024-09-05T07:36:36Z"
  mapUDFCount: 1
  message: Pipeline paused
  observedGeneration: 5
  phase: Paused
  reduceUDFCount: 0
  sinkCount: 1
  sourceCount: 1
  udfCount: 1
  vertexCount: 3
skohli@macos-JQWR9T560R numaflow % kubectl get pl                        
NAME              PHASE    VERTICES   AGE   MESSAGE
simple-pipeline   Paused   3          8h    Pipeline paused

```

**Comment by kohlisid** [comment]
```
  lastUpdated: "2024-09-05T15:36:32Z"
  mapUDFCount: 1
  observedGeneration: 6
  phase: Running
  reduceUDFCount: 0
  sinkCount: 1
  sourceCount: 1
  udfCount: 1
  vertexCount: 3
skohli@macos-JQWR9T560R numaflow % kubectl get pl                        
NAME              PHASE     VERTICES   AGE   MESSAGE
simple-pipeline   Running   3          8h   
```

**Comment by kohlisid** [comment]
```
skohli@macos-JQWR9T560R numaflow %   kubectl patch pl simple-pipeline --type=merge --patch '{"spec": {"lifecycle": {"desiredPhase": "Paused"}}}' 

pipeline.numaflow.numaproj.io/simple-pipeline patched (no change)
skohli@macos-JQWR9T560R numaflow % kubectl get pl                                                                                               
NAME              PHASE    VERTICES   AGE   MESSAGE
simple-pipeline   Failed   3          11s   ISB Service not found.

```

**Comment by kohlisid** [comment]
```
skohli@macos-JQWR9T560R numaflow %   kubectl patch pl simple-pipeline --type=merge --patch '{"spec": {"lifecycle": {"desiredPhase": "Paused"}}}'

pipeline.numaflow.numaproj.io/simple-pipeline patched
skohli@macos-JQWR9T560R numaflow % kubectl get po                                                                                               
NAME                               READY   STATUS      RESTARTS   AGE
simple-pipeline-cln-ac4416-l6xds   0/1     Completed   0          24s
skohli@macos-JQWR9T560R numaflow % kubectl get pl
NAME              PHASE    VERTICES   AGE   MESSAGE
simple-pipeline   Failed   3          16s   ISB Service not found.

```

**Comment by kohlisid** [comment]
```
  drainedOnPause: false
  lastUpdated: "2024-09-05T16:48:56Z"
  mapUDFCount: 1
  observedGeneration: 4
  phase: Running
  reduceUDFCount: 0
  sinkCount: 1
  sourceCount: 1
  udfCount: 1
  vertexCount: 3
skohli@macos-JQWR9T560R numaflow % kubectl get pl                        
NAME              PHASE     VERTICES   AGE    MESSAGE
simple-pipeline   Running   3          6m9s  
```

File: pkg/reconciler/pipeline/controller.go [hunk 1]
```diff
@@ -599,7 +620,8 @@ func buildVertices(pl *dfv1.Pipeline) map[string]dfv1.Vertex {
 		copyVertexTemplate(pl, vCopy)
 		copyVertexLimits(pl, vCopy)
 		replicas := int32(1)
-		if pl.Status.Phase == dfv1.PipelinePhasePaused {
+		// If the desired phase is pause or we are in the middle of pausing we should not start any vertex replicas
+		if isLifecycleChange(pl) {
```
Comment (juliev0) [inline pkg/reconciler/pipeline/controller.go:624]: So, I just noticed that `isLifecycleChange() == true` in the case that the oldPhase was Paused but the desiredPhase==Running. I haven't read enough of the other code yet to confirm that that's okay here ?

Comment (kohlisid) [reply pkg/reconciler/pipeline/controller.go:624]: That should be a `resume` pipeline case and I kept it like that for it to be handled by the 
[resume-logic](https://github.com/kohlisid/numaflow/blob/e7c32c1857a448b3399721062ff619208917ec91/pkg/reconciler/pipeline/controller.go#L796)

File: pkg/reconciler/pipeline/controller.go [hunk 2]
```diff
@@ -144,9 +144,17 @@ func (r *pipelineReconciler) reconcile(ctx context.Context, pl *dfv1.Pipeline) (
 	}()
 
 	pl.Status.SetObservedGeneration(pl.Generation)
-
-	if oldPhase := pl.Status.Phase; pl.Spec.Lifecycle.GetDesiredPhase() == dfv1.PipelinePhasePaused ||
-		oldPhase == dfv1.PipelinePhasePaused || oldPhase == dfv1.PipelinePhasePausing {
+	// Regular pipeline change
+	// This should be happening in call cases to ensure a clean initialization regardless of the lifecycle phase
```
Comment (juliev0) [inline pkg/reconciler/pipeline/controller.go:148]:
```suggestion
	// This should be happening in all cases to ensure a clean initialization regardless of the lifecycle phase
```

Comment (juliev0) [reply pkg/reconciler/pipeline/controller.go:148]: I assume that's what you meant?

**Review by whynowy — APPROVED** [review]

File: pkg/reconciler/pipeline/controller.go (hunk 2 above)
Comment (whynowy) [reply pkg/reconciler/pipeline/controller.go:148]: Merging this, will address it in another PR.

**Pull request merged**